py-polars/src/series/buffers.rs

```diff
@@ -281,3 +277,123 @@ fn get_boolean_buffer_length_in_bytes(length: usize, offset: usize) -> usize {
         n_bytes + 1
     }
 }
+
+#[pymethods]
+impl PySeries {
+    /// Construct a PySeries from information about its underlying buffers.
+    #[staticmethod]
+    unsafe fn _from_buffers(
+        dtype: Wrap<DataType>,
+        data: PySeries,
+        validity: Option<PySeries>,
+        offsets: Option<PySeries>,
+    ) -> PyResult<Self> {
+        let dtype = dtype.0;
+        let data = data.series;
+
+        if validity.is_none() && offsets.is_none() {
+            let s = data.strict_cast(&dtype).map_err(PyPolarsErr::from)?;
+            return Ok(s.into());
+        }
+
+        let validity = match validity {
+            Some(s) => {
+                let dtype = s.series.dtype();
+                if !dtype.is_bool() {
+                    return Err(PyTypeError::new_err(format!(
+                        "validity buffer must have data type Boolean, got {:?}",
+                        dtype
+                    )));
+                }
+                Some(series_to_bitmap(s.series).unwrap())
+            },
+            None => None,
+        };
+
+        let s = match dtype.to_physical() {
+            dt if dt.is_numeric() => {
+                with_match_physical_numeric_polars_type!(dt, |$T| {
+                    let data = series_to_buffer::<$T>(data);
+                    from_buffers_num_impl::<<$T as PolarsNumericType>::Native>(data, validity)?
+                })
+            },
+            DataType::Boolean => {
+                let data = series_to_bitmap(data)?;
+                from_buffers_bool_impl(data, validity)?
+            },
+            DataType::String => {
+                let offsets = match offsets {
+                    Some(s) => {
+                        let dtype = s.series.dtype();
+                        if !matches!(dtype, DataType::Int64) {
+                            return Err(PyTypeError::new_err(format!(
+                                "offsets buffer must have data type Int64, got {:?}",
+                                dtype
+                            )));
+                        }
+                        series_to_offsets(s.series)
+                    },
+                    None => return Err(PyTypeError::new_err(
+                        "`from_buffers` cannot create a String column without an offsets buffer",
+                    )),
+                };
+                let data = series_to_buffer::<UInt8Type>(data);
+                from_buffers_string_impl(data, validity, offsets)?
+            },
+            dt => {
+                return Err(PyTypeError::new_err(format!(
+                    "`from_buffers` not implemented for `dtype` {dt}",
+                )))
+            },
+        };
+
+        let out = s.strict_cast(&dtype).map_err(PyPolarsErr::from)?;
+        Ok(out.into())
+    }
+}
+
+fn series_to_buffer<T>(s: Series) -> Buffer<T::Native>
+where
+    T: PolarsNumericType,
+{
+    let ca: &ChunkedArray<T> = s.as_ref().as_ref();
```

Comment:
This is magic to me but it seems to work 😄

Reply:
See the `AsRef` ipl on `Series` or `SeriesTrait`. ;)

Reply:
Nice, makes sense now!